src/afa/afa.cc

```diff
@@ -1003,41 +1002,130 @@ void Mata::Afa::construct(
-	StringToStateMap*                    state_map)
+Afa Mata::Afa::construct(
+        const Mata::IntermediateAut&         inter_aut,
+        Alphabet*                            alphabet,
+        StringToStateMap*                    state_map)
 { // {{{
-	assert(nullptr != aut);
+    Afa aut;
+    assert(nullptr != alphabet);
 
-	bool remove_symbol_map = false;
-	if (nullptr == symbol_map)
-	{
-		symbol_map = new StringToSymbolMap();
-		remove_symbol_map = true;
-	}
+    if (!inter_aut.is_afa()) {
+        throw std::runtime_error(std::string(__FUNCTION__) + ": expecting type \"" +
+                                 Mata::Afa::TYPE_AFA + "\"");
+    }
 
-	auto release_res = [&](){ if (remove_symbol_map) delete symbol_map; };
+    bool remove_state_map = false;
+    if (nullptr == state_map) {
+        state_map = new StringToStateMap();
+        remove_state_map = true;
+    }
 
-  Mata::Nfa::OnTheFlyAlphabet alphabet(*symbol_map);
+    // a lambda for translating state names to identifiers
+    auto get_state_name = [&state_map, &aut](const std::string& str) {
+        if (!state_map->count(str)) {
+            State state = aut.add_new_state();
+            state_map->insert({str, state});
+            return state;
+        } else {
+            return (*state_map)[str];
+        }
+    };
+
+    // a lambda for cleanup
+    auto clean_up = [&]() {
+        if (remove_state_map) { delete state_map; }
+    };
+
+    // lambda returning true if node is operator of given type
+    auto is_node_operator =
+            [](const FormulaNode& node, FormulaNode::OperatorType type) -> bool {
+        return node.is_operator() && node.operator_type == type;
+    };
+
+    // lambda creates Node from set of strings which are state names
+    auto create_node = [get_state_name](
+            const std::unordered_set<std::string>& states_names) -> Node {
+        Node tgt_node;
+        for (const std::string& s : states_names)
+        {
+            tgt_node.insert(get_state_name(s));
+        }
+
+        return tgt_node;
+    };
+
+    for (const auto& str : inter_aut.initial_formula.collect_node_names())
+    {
+        State state = get_state_name(str);
+        aut.initialstates.insert(state);
+    }
 
-	try
-	{
-		construct(aut, parsec, &alphabet, state_map);
-	}
-	catch (std::exception&)
-	{
-		release_res();
-		throw;
-	}
+    for (const auto& str : inter_aut.final_formula.collect_node_names())
+    {
+        State state = get_state_name(str);
+        aut.finalstates.insert(state);
+    }
 
-	release_res();
-} // construct(StringToSymbolMap) }}}
+    for (const auto& trans : inter_aut.transitions)
+    {
+        State src_state = get_state_name(trans.first.name);
+        if (trans.second.node.is_operand() && trans.second.node.operand_type == FormulaNode::SYMBOL)
+        {
+            Symbol symbol = alphabet->translate_symb(trans.second.node.name);
+            aut.add_trans(src_state, symbol, Node());
+            continue;
+        }
```

Comment:
I'm not sure if I understand this part correctly. I guess that this condition is satisfied as soon as a transition like `q1 a` is processed (or `q1 a &`, unfortunately, I did not find an example so I am not sure), which means that the destination nodes are omitted.

In my implementation of `AFAs`, once I reach the empty node `Node()`, I can visit any node by reading any symbol. Therefore, I accept anything after I reach the empty node `Node()` since the empty node is always final (even if the set of final states is an empty set).

Am I right that this is also an intended semantics of transitions like `q1 a`? If I read `a` in the state `q1`, I accept anything. If so, ok. (I just totally forgot that such transitions could exist so I have to fix my inverse transition relation which does not support that.)

Reply:
Yes, this take care about transition `q1 a`. I didn't support it either in the first version of conversion. However, then I received an example of AFA by @jurajsic where such transitions exist. So we need to take care about such transition. There are also transitions without symbol, that is `q1 q2`.

Reply:
Ok, thanks for your answer. You have mentioned a transition `q1 q2`, am I right that it is an epsilon transition which will be implemented at some point later?

Reply:
yes, it is an epsilon transition.